File: mac/LRActionKit/LRActionKit/ActionKitSingleton.m
#import "ActionKitSingleton.h"
#import "LRActionKit-Swift.h"


@implementation ActionKitSingleton

static ActionKitSingleton *sharedActionKit;

+ (instancetype)sharedActionKit {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedActionKit = [ActionKitSingleton new];
    });
    return sharedActionKit;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _optionRegistry = [OptionRegistry new];
        [_optionRegistry addOptionType:[CheckboxOptionType new]];
        [_optionRegistry addOptionType:[MultipleChoiceOptionType new]];
        [_optionRegistry addOptionType:[TextOptionType new]];
    }
    return self;
}

@end

File: LRActionKit/Source/ActionKitSingleton.swift
import Foundation
import PackageManagerKit
import ExpressiveCasting

@objc
public final class ActionKitSingleton: NSObject {

    public static let sharedActionKit = ActionKitSingleton()

    public let optionRegistry: OptionRegistry = ({
        let or = OptionRegistry()
        or.addOptionType(CheckboxOptionType())
        or.addOptionType(MultipleChoiceOptionType())
        or.addOptionType(TextOptionType())
        return or
    })()

    public var packageManager: LRPackageManager!

    public var postMessageBlock: (message: JSONObject, completionBlock: (error: NSError?, response: JSONObject?) -> Void) -> Void

    private override init() {
        postMessageBlock = { message, completionBlock in
            fatalError("postMessageBlock not set, sending message \(message)")
        }

        super.init()
    }

    public func postMessage(message: JSONObject, completionBlock: (error: NSError?, response: JSONObject?) -> Void) {
        postMessageBlock(message: message, completionBlock: completionBlock)
    }

}
